src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs

```diff
@@ -11,14 +11,16 @@
 
 namespace System.Xml
 {
-    internal sealed class XmlDictionaryAsyncCheckWriter : XmlDictionaryWriter
+    internal sealed class XmlDictionaryAsyncCheckWriter : XmlDictionaryWriter, IXmlTextWriterInitializer
     {
         private readonly XmlDictionaryWriter _coreWriter;
+        private readonly IXmlTextWriterInitializer? _initializer;
         private Task? _lastTask;
 
         public XmlDictionaryAsyncCheckWriter(XmlDictionaryWriter writer)
         {
             _coreWriter = writer;
+            _initializer = writer as IXmlTextWriterInitializer;
```

Comment:
Does this need to be stored separately?

Couldn't SetOutput instead just be:
```C#
public void SetOutput(Stream stream, Encoding encoding, bool ownsStream)
{
    CheckAsync();
    if (_coreWriter is IXmlTextWriterInitializer initializer)
    {
        initializer.SetOutput(stream, encoding, ownsStream);
    }
}
```
?  Or is there a reason we want to avoid that cast in SetOutput?

Reply:
Honestly, that's just the way I did it and I hadn't considered this alternative. The whole idea of this change is to bring a performance boost back that was lost when XmlDictionaryAsyncCheckWriter was introduced which allows WCF to reuse the writer instead of having to new up new ones all the time.

From a perf perspective we are saving lots of memory allocations by enabling this, so I'm not sure one more field is too detrimental. Otoh, cast operations aren't terribly expensive either, and I do like the pattern matching cast pattern. But I suppose how much I like it depends on how hot this 'SetOutput' path is in WCF. @mconnew?

Reply:
SetOutput will be called once for every incoming request. The amount of work done per request will dwarf the cost of the cast by multiple orders of magnitude. The main thing is to avoid the allocation and GC cleanup which either will do. SetOutput is already going to be a virtual function call and with the wrapping, we do it twice, once in the call from WCF, and again in the wrapped call, so I don't think it matters much the cost of the cast. My preference in this case is err on the side of code readability.